### PR TITLE
some restructuring to support local directories

### PR DESCRIPTION
**Pull request by stephchen**
Description
-----------

splits the concept of `project_spec.dir` into a dir where the code to run lives (either pulled from remote or pointing to the user's local dir) and then an auxiliary dir to load config files, etc. also adds a LaunchSource enum which idk how necessary this is but it feels cleaner than relying on regexes over and over again.

Testing
-------

added a test

Release Notes
-------------

Below, please enter user-facing release notes as one or more bullet points. If your change is not user-visible, write `NO RELEASE NOTES` instead, with no bullet points.

------------- BEGIN RELEASE NOTES ------------------

------------- END RELEASE NOTES --------------------

### CODE REVIEW

File: wandb/sdk/launch/_project_spec.py [hunk 1]
```diff
@@ -54,8 +61,20 @@ def __init__(
         if "entry_point" in overrides:
             self.add_entry_point(overrides["entry_point"])
 
+        if utils._is_wandb_uri(self.uri):
+            self.source = LaunchSource.WANDB
+            self.project_dir = tempfile.mkdtemp()
+        elif utils._is_github_uri(self.uri):
+            self.source = LaunchSource.GITHUB
+            self.project_dir = tempfile.mkdtemp()
+        else:
```
Comment (KyleGoyette) [inline wandb/sdk/launch/_project_spec.py:70]: We should have a regex check to see if it is indeed a local directory, and error out if now

File: wandb/sdk/launch/_project_spec.py [hunk 2]
```diff
@@ -248,8 +268,13 @@ def create_project_from_spec(run_spec: Dict[str, Any], api: Api) -> Project:
...
 def fetch_and_validate_project(project: Project, api: Api) -> Project:
-    project._fetch_project_local(api=api)
-    project._copy_config_local()
+    if project.source == LaunchSource.LOCAL:
+        if not project._entry_points:
+            wandb.termlog("Entry point for repo not specified, defaulting to main.py")
+            project.add_entry_point("main.py")
+    else:
+        project._fetch_project_local(api=api)
+    project._copy_config_local()  # copy config into self.dir even if we're local
```
Comment (KyleGoyette) [inline wandb/sdk/launch/_project_spec.py:277]: this is copying into the aux_dir now right?

File: wandb/sdk/launch/_project_spec.py [hunk 3]
```diff
@@ -54,8 +61,20 @@ def __init__(
...
+        else:
+            # assume local
+            self.source = LaunchSource.LOCAL
+            self.project_dir = self.uri
+
+        self.aux_dir = tempfile.mkdtemp()
+
         self.run_id = generate_id()
-        self.dir = tempfile.mkdtemp()
         self.config_path = DEFAULT_CONFIG_PATH
```
Comment (KyleGoyette) [inline wandb/sdk/launch/_project_spec.py:78]: I think the config path should be `os.path.join(self.aux_dir, DEFAULT_CONFIG_PATH)`

**Review by KyleGoyette — CHANGES_REQUESTED** [review]
A few changes I think, but this is a good direction

File: wandb/sdk/launch/_project_spec.py (hunk 3 above)
Comment (stephchen) [reply wandb/sdk/launch/_project_spec.py:78]: oh yeah it does this in the copy fn but looks like the docker side assumes `config_path` is the full, good catch

**Review by KyleGoyette — APPROVED** [review]
nice

File: wandb/sdk/launch/utils.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 
 # TODO: this should be restricted to just Git repos and not S3 and stuff like that
-_GIT_URI_REGEX = re.compile(r"^[^/]*:")
+_GIT_URI_REGEX = re.compile(r"^https://github\.com")
```
Comment (KyleGoyette) [inline wandb/sdk/launch/utils.py:15]: Oh wait, we should support different git repo providers besides github.

**Pull request merged**